cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,329 @@
+	}
+}
+
+// installPluginFromZip extracts a plugin zip file, validates and
+// installs the plugin
+func installPluginFromZip(ctx context.Context, zipPath string, force bool) error {
+	logger := log.GetLogger(ctx)
+	archive, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer archive.Close()
+	// require one and only one file with name in the format
+	// `notation-{plugin-name}`
+	for _, f := range archive.File {
+		if !f.Mode().IsRegular() || strings.Contains(f.Name, "..") {
+			continue
+		}
+		// validate and get plugin name from file name
+		pluginName, err := notationplugin.ExtractPluginNameFromFileName(f.Name)
+		if err != nil {
+			logger.Infof("processing file %s. File name not in format notation-{plugin-name}, skipped", f.Name)
+			continue
+		}
+		fileInArchive, err := f.Open()
+		if err != nil {
+			return err
+		}
+		defer fileInArchive.Close()
+		return installPluginExecutable(ctx, f.Name, pluginName, fileInArchive, force)
+	}
+	return fmt.Errorf("no valid plugin file was found in %s. Plugin file name must in format notation-{plugin-name}", zipPath)
+}
+
+// installPluginFromTarGz extracts and untar a plugin tar.gz file, validates and
+// installs the plugin
+func installPluginFromTarGz(ctx context.Context, tarGzPath string, force bool) error {
+	logger := log.GetLogger(ctx)
+	rc, err := os.Open(tarGzPath)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	decompressedStream, err := gzip.NewReader(rc)
+	if err != nil {
+		return err
+	}
+	defer decompressedStream.Close()
+	tarReader := tar.NewReader(decompressedStream)
+	// require one and only one file with name in the format
+	// `notation-{plugin-name}`
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if !header.FileInfo().Mode().IsRegular() || strings.Contains(header.Name, "..") {
+			continue
+		}
+		// validate and get plugin name from file name
+		pluginName, err := notationplugin.ExtractPluginNameFromFileName(header.Name)
+		if err != nil {
+			logger.Infof("processing file %s. File name not in format notation-{plugin-name}, skipped", header.Name)
+			continue
+		}
+		return installPluginExecutable(ctx, header.Name, pluginName, tarReader, force)
+	}
+	return fmt.Errorf("no valid plugin file was found in %s. Plugin file name must in format notation-{plugin-name}", tarGzPath)
+}
+
+// installPluginExecutable extracts, validates, and installs a plugin file
+func installPluginExecutable(ctx context.Context, fileName string, pluginName string, fileReader io.Reader, force bool) error {
+	// sanity check
+	if runtime.GOOS == "windows" && filepath.Ext(fileName) != ".exe" {
+		return fmt.Errorf("on Windows, plugin executable file %s is missing the '.exe' extension", fileName)
+	}
+	if runtime.GOOS != "windows" && filepath.Ext(fileName) == ".exe" {
+		return fmt.Errorf("on %s, plugin executable file %s cannot have the '.exe' extension", runtime.GOOS, fileName)
+	}
```

Comment:
nit: should you use [build constraints](https://pkg.go.dev/cmd/go#hdr-Build_constraints) for those checks?

Reply:
Refactored this part to notation-go.